src/hugr.rs

```diff
@@ -70,6 +75,193 @@ pub struct Wire(Node, usize);
 
 /// Public API for HUGRs.
 impl Hugr {
+    /// Apply a simple replacement operation to the HUGR.
+    pub fn apply_simple_replacement(
+        &mut self,
+        r: SimpleReplacement,
+    ) -> Result<(), SimpleReplacementError> {
+        // 1. Check the parent node exists and is a DFG node.
+        if self.get_optype(r.parent).tag() != OpTag::Dfg {
+            return Err(SimpleReplacementError::InvalidParentNode());
+        }
+        // 2. Check that all the to-be-removed nodes are children of it and are leaves.
+        for node in &r.removal {
+            if self.hierarchy.is_root(node.index)
+                || self.hierarchy.parent(node.index).unwrap() != r.parent.index
+                || self.hierarchy.has_children(node.index)
+            {
+                return Err(SimpleReplacementError::InvalidRemovedNode());
+            }
+        }
+        // 3. Do the replacement.
+        // 3.1. Add copies of all replacement nodes and edges to self. Exclude Input/Output nodes.
+        // Create map from old NodeIndex (in r.replacement) to new NodeIndex (in self).
+        let mut index_map: HashMap<NodeIndex, NodeIndex> = HashMap::new();
+        let replacement_nodes = r
+            .replacement
+            .hierarchy
+            .children(r.replacement.root().index)
+            .map_into::<Node>()
+            .collect::<Vec<Node>>();
+        // number of replacement nodes including Input and Output:
+        let replacement_sz = replacement_nodes.len();
+        // slice of nodes omitting Input and Output:
+        let replacement_inner_nodes = &replacement_nodes[1..replacement_sz - 1];
+        for &node in replacement_inner_nodes {
+            // Check there are no const inputs.
+            if !r
+                .replacement
+                .get_optype(node)
+                .signature()
+                .const_input
+                .is_empty()
+            {
+                return Err(SimpleReplacementError::InvalidReplacementNode());
+            }
+        }
+        let self_input_node_index = self.hierarchy.first(r.parent.index).unwrap();
+        let replacement_output_node = replacement_nodes[replacement_sz - 1];
```

Comment:
Nitpicky, but it's slightly clearer
```suggestion
        let replacement_output_node = *replacement_nodes.last().unwrap();
```